contributors/design-proposals/node/node-usernamespace-remapping.md

```diff
@@ -48,134 +49,286 @@ Enable user namespace support in a kubernetes cluster so that workloads that wor
+```golang
+        // Use host's user namespace for the pod.
+        // This field is alpha-level and can be set to true/false only if HostUserNamespace feature-gate is enabled.
+        // Optional: Defaults to nil which means behavior will be container runtime defined.
+        // +k8s:conversion-gen=false
+        // +optional
+        HostUserNamespace *bool `json:"hostUserNamespace,omitempty" protobuf:"varint,31,opt,name=hostUserNamespace"`
+```
+Today end users are NOT provided any guaranteed behavior by the Kubernetes related to usernamespace remapping support at runtime. If runtime has usernamespace remapping disabled, pods will run otherwise pod containers will fail to start. In other words, behavior is (container runtime) configuration defined. `nil` value of `HostUserNamespace` will show the same behavior i.e configuration(runtime and the kubelet flags `--userns-{uid/gid}map` ) defined usernamespace behavior. If the kubelet flags are to be set and `HostUserNamespace` value will be changed from `nil` to `false` otherwise `true`.  
+However, with an added improvement of running pods with remapped usernamespace "when possible". "when possible" means configuration at runtime and kubelet is usernamespace remapping enabled and pod does not specify any configuration which is not possible to provide in remapped usernamespace. 
 
-- **Beta:** Explicit API to request host user-namespace in pod spec
-          This is being targeted under Beta graduation plans.
+`HostUserNamespace` field will be gated behind `HostUserNamespace` feature-gate. To set it to `true` or `false`, feature-gate will have to be enabled. If feature-gate not enabled or feature-gate is enabled but any value is not set, `nil` will be the default value. 
 
-### CRI API Changes
-Proposed CRI API changes:
+#### New Field in the Volume api object
 
+`HostUserNamespace` field will be introduced in the `Volume`
 ```golang
-// Runtime service defines the public APIs for remote container runtimes
-service RuntimeService {
+// Volume represents a named volume in a pod that may be accessed by any container in the pod. 
+type Volume struct {
+        // Volume's name.
+        // Must be a DNS_LABEL and unique within the pod. 
+        // More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+        Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+        // VolumeSource represents the location and type of the mounted volume.
+        // If not specified, the Volume is implied to be an EmptyDir.
+        // This implied behavior is deprecated and will be removed in a future version.
+        VolumeSource `json:",inline" protobuf:"bytes,2,opt,name=volumeSource"`
+        // Optional: Defaults to value of hostUserNamespace field in the pod spec
+        // +k8s:conversion-gen=false
+        // +optional
+        HostUserNamespace *bool `json:"hostUserNamespace,omitempty" protobuf:"varint,3,opt,
```

Comment:
This troubles me. I think this might be the first feature in K8s where two separate resources will have to coordinate well with one another in order to work. Forcing the user to think about namespaces at this level is a recipe for trouble, IMO. Also, how will K8s handle node failures and premature detaches? If these volumes get chowned, before getting unchowned. Users volumes will be in a broken state.

It would almost be better to fail (i.e. not handle the case of volumes yet).